src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -2436,118 +2501,328 @@ public void UpdateIsNullableEnabledForConstructorsAndFields(bool useStatic, bool
+                        else
+                        {
+                            groupsBuilder.Add(declaredMembers.InstanceInitializers[i]);
+                        }
+                    }
+
+                    return groupsBuilder.ToImmutableAndFree();
                 }
+            }
 
-                initializers.Free();
-                return builder.ToImmutableAndFree();
+            public void AddInstanceInitializerForPositionalMembers(FieldOrPropertyInitializer initializer)
+            {
+                if (InstanceInitializersForPositionalMembers is null)
+                {
+                    InstanceInitializersForPositionalMembers = ArrayBuilder<FieldOrPropertyInitializer>.GetInstance();
+                }
+
+                InstanceInitializersForPositionalMembers.Add(initializer);
             }
 
-            private static ImmutableArray<FieldOrPropertyInitializer> ToReadonlyAndFree(ArrayBuilder<FieldOrPropertyInitializer.Builder> group, ref int syntaxLength)
+            public IReadOnlyCollection<Symbol> GetNonTypeNonIndexerMembers(DeclaredMembersAndInitializers declaredMembers)
             {
-                if (group.Count == 0)
+                return NonTypeNonIndexerMembers ?? (IReadOnlyCollection<Symbol>)declaredMembers.NonTypeNonIndexerMembers;
+            }
+
+            public void AddNonTypeNonIndexMember(Symbol member, DeclaredMembersAndInitializers declaredMembers)
+            {
+                if (NonTypeNonIndexerMembers is null)
                 {
-                    group.Free();
-                    return ImmutableArray<FieldOrPropertyInitializer>.Empty;
+                    NonTypeNonIndexerMembers = ArrayBuilder<Symbol>.GetInstance(declaredMembers.NonTypeNonIndexerMembers.Length + 1);
+                    NonTypeNonIndexerMembers.AddRange(declaredMembers.NonTypeNonIndexerMembers);
                 }
 
-                var builder = ArrayBuilder<FieldOrPropertyInitializer>.GetInstance(group.Count);
+                NonTypeNonIndexerMembers.Add(member);
+            }
+
+            public void UpdateIsNullableEnabledForConstructorsAndFields(bool useStatic, CSharpCompilation compilation, CSharpSyntaxNode syntax)
+            {
+                ref bool isNullableEnabled = ref GetIsNullableEnabledForConstructorsAndFields(useStatic);
+                isNullableEnabled = isNullableEnabled || compilation.IsNullableAnalysisEnabledIn(syntax);
+            }
+
+            private ref bool GetIsNullableEnabledForConstructorsAndFields(bool useStatic)
+            {
+                return ref useStatic ? ref IsNullableEnabledForStaticConstructorsAndFields : ref IsNullableEnabledForInstanceConstructorsAndFields;
+            }
 
-                foreach (FieldOrPropertyInitializer.Builder initializer in group)
+            internal static ImmutableArray<ImmutableArray<FieldOrPropertyInitializer>> ToReadOnlyAndFree(ArrayBuilder<ArrayBuilder<FieldOrPropertyInitializer>> initializers)
```

Comment:
Is this method called from `DeclaredMembersAndInitializersBuilder` only? If so, please consider moving it there.

Reply:
I thought of making these local functions, but wanted to minimize code churn (less moving code around). I'll leave as-is.

---
In reply to: [564117116](https://github.com/dotnet/roslyn/pull/50650#discussion_r564117116) [](ancestors = 564117116)